### ISSUE TITLE
support .net image slaves

### ISSUE THREAD

**Issue by tmds** [openshift/jenkins#626]
@bparees, hi, I was going to add .NET Core 2.1 slaves, but I noticed there are no more .NET Core slaves.

These were removed in https://github.com/openshift/jenkins/pull/431 where I got a mention (which I can't remember).

> @tmds fyi we're removing the pre-configured .net slave image configuration after realizing that backwards compatibility is virtually impossible to guarantee since the .net image includes the oc client binary, which must be aligned with the cluster version the slave image is used on.

I'm not sure why our images would include the `oc` client? But probably we can get rid of it.

CC @aslicerh @omajid

**Comment by aslicerh** [comment]
Can someone clarify what is meant by `oc client binary`? Do you mean `atomic-openshift-clients...`?

**Comment by tmds** [comment]
> Can someone clarify what is meant by oc client binary? Do you mean atomic-openshift-clients...?

I guess so:
```
$ yum whatprovides /usr/bin/oc
atomic-openshift-clients-3.9.30-1.git.0.dec1ba7.el7.x86_64 : Origin Client binaries for Linux
Repo        : @rhel-server-ose-rpms
Matched from:
Filename    : /usr/bin/oc
```

The nodejs slave has the `oc` binaries too. Aren't slave supposed to have it, to be able to interact with the cluster?

**Comment by aslicerh** [comment]
>  Aren't slave supposed to have it, to be able to interact with the cluster?

Yeah, that's why I asked. That package is coming from the base slave image, not us directly.

**Comment by bparees** [comment]
yes, the slave base image has it because it's commonly needed to execute jobs (the openshift client plugin depends on it).

**Comment by tmds** [comment]
@bparees is the reason for removing the .NET Core streams that the 'oc' version doesn't match the cluster version? How can we solve that? Do we need to include the .NET Core slaves in this repo?

**Comment by bparees** [comment]
like i said in the PR, the issue here is that we need to version-align the slave images w/ the master image and the cluster version (to guarantee the oc binary inside the slave image is compatible w/ the cluster the slave images is running on).

For "our" slave images (nodejs, maven) we publish version specific tags of them (xxx:v3.7, xxx:v3.9, etc) so we have a specific tag for the slave image that we can configure the master to use.  So we know a v3.9 cluster will use the v3.9 master and we know the v3.9 master will use the v3.9 slaves, and we know the v3.9 slaves will contain a v3.9 oc binary which can talk to the cluster.

For images we don't control (namely, your .NET image) we can't ensure you guys publish version specific tags that match openshift versions (you're not part of our release process).  So don't have a way to configure the v3.9 jenkins master image to point to a known-compatible .NET slave image tag that it should be using that contains the v3.9 oc binary.  And we don't expect you to deal w/ that either.

Therefore the path of least resistance was to stop including the configuration for the .NET slave in our jenkins master, and simply force users who want to use the .NET slave to pick an appropriate .NET slave themselves which will work w/ their cluster version.

On top of the logistical/tagging challenges that lead to this, I also think it was a mistake for us to try to provide too much "default" configuration in the jenkins master.  It sets the wrong expectations for users.  The configuration we provide out of the box should be considered as "examples/demos" and not "production ready", so the less configuration we provide (and thus putting the burden on users to understand how to configure the image for their specific needs instead of expecting the out of box configuration will "just work" for all cases, when it's very difficult/impossible for us to guarantee that), the better.

I think the appropriate solution is to simply make it more obvious to users what other slaves images exist and how to add them to their jenkins configuration.  We already have a fair bit of documentation about how to add slaves:
https://docs.openshift.org/latest/using_images/other_images/jenkins.html#configuring-the-jenkins-kubernetes-plug-in

and i think we'd be happy to add the .NET image information to this doc:
https://docs.openshift.org/latest/using_images/other_images/jenkins_slaves.html

**Comment by tmds** [comment]
> For "our" slave images (nodejs, maven) we publish version specific tags of them (xxx:v3.7, xxx:v3.9, etc) so we have a specific tag for the slave image that we can configure the master to use. So we know a v3.9 cluster will use the v3.9 master and we know the v3.9 master will use the v3.9 slaves, and we know the v3.9 slaves will contain a v3.9 oc binary which can talk to the cluster.

Should we have similar tags for the .NET Core slave, so we also have an oc binary that can talk to the cluster?

**Comment by aslicerh** [comment]
> On top of the logistical/tagging challenges that lead to this, I also think it was a mistake for us to try to provide too much "default" configuration in the jenkins master.

Does this mean that you intend to also stop defaulting nodejs and maven? What is the long term plan for integrating things by 'default'?

**Comment by bparees** [comment]
> Should we have similar tags for the .NET Core slave, so we also have an oc binary that can talk to the cluster? 

it would be a good idea, but again we didn't want to try to enforce that everyone else creating slave images produce openshift versioned tags.

> Can we use tags on the base image for this?

yes, you should be able to base your slave image on our slave base:v3.x since that'll get you the oc binary for "x".

**Comment by bparees** [comment]
> Does this mean that you intend to also stop defaulting nodejs and maven? What is the long term plan for integrating things by 'default'?

i think we're stuck including those since people have come to expect it (and because we do want that example/demo experience to work out of the box), we just don't want to add more, and we're more comfortable including config for things that are under our control (we do automated testing around those images as part of our openshift build/release process, for example).

**Comment by aslicerh** [comment]
> it would be a good idea, but again we didn't want to try to enforce that everyone else creating slave images produce openshift versioned tags.

Is there any defined standard/convention on what these tags should be formatted as?

**Comment by tmds** [comment]
> Is there any defined standard/convention on what these tags should be formatted as?

`v3.x` tags based on the corresponding `v3.x` slave base image tags.

**Comment by aslicerh** [comment]
> v3.x tags based on the corresponding v3.x slave base image tags.

Right, but we also have our own versioning (2.0-14, etc) that we need to make note of, and I assume that anything else derived from the base image would as well, so I was wondering if there was any expectation here of how those merge. Also, I was curious if the tooling in general expected anything.

**Comment by tmds** [comment]
The (# .NET Core Versions) x (# OpenShift Versions) means a lot of work. We need to look at what we gain for that work.

Looking at https://access.redhat.com/support/policy/updates/openshift, it seems all minor 3.x versions are in Full Support till 2020. @bparees is that right?

**Comment by bparees** [comment]
> Is there any defined standard/convention on what these tags should be formatted as?

v3.9, v3.10, v3.11, etc:
https://hub.docker.com/r/openshift/jenkins-slave-base-centos7/tags/
https://access.redhat.com/containers/?tab=tags#/registry.access.redhat.com/openshift3/jenkins-slave-base-rhel7

> The (# .NET Core Versions) x (# OpenShift Versions) means a lot of work. We need to look at what we gain for that work.

yeah i would not expect you to do that, i'd just provide all .NET versions you want to support, for the latest openshift version.  When a new openshift version comes out, provide .NET versions you want to support at that time (meaning you might EOL some older versions and introduce new ones only for that new openshift version.  That is the pattern we are following w/ nodejs and maven).

> Looking at https://access.redhat.com/support/policy/updates/openshift, it seems all minor 3.x versions are in Full Support till 2020. @bparees is that right?

I think that is currently accurate, but up until v3.6 we only shipped a single version because the 3.6 oc binary was compatible with 3.0->3.6.  It was with the introduction of 3.7 that the binary was not fully backwards compatible and we needed to start getting version specific.
